File: cpp/2300-successful-pairs-of-spells-and-potions.cpp
// Time: O(N * logN)
// Space: O(N)

class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        sort(potions.begin(), potions.end());

        int n = spells.size();
        int m = potions.size();
        vector<int> pairs(n);

        for(int i = 0; i < n; i++) {
            int spell = spells[i];

            int start = 0, end = m - 1;
            int curr;

            while(start <= end) {
                curr = start + (end-start)/2;
                long long strength = (long long)potions[curr] * (long long)spell;
                if(strength < success) {
                    start = curr + 1;
                }
                else {
                    end = curr - 1;
                }
            }
            pairs[i] = m - start;
        }

        return pairs;
    }
};

File: python/2300-successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        pairs = []
        potions.sort()
        n = len(potions)

        for i in range(len(spells)):
            l, r = 0, len(potions) - 1
            
            while l <= r:
                m = (l + r) // 2
                if spells[i] * potions[m] >= success:
                    r = m - 1
                else:
                    l = m + 1
            if l < len(potions):
                pairs.append(n - l)
            else:
                pairs.append(0)

        return pairs
        
        
